### PR TITLE
Migrate `fs-api` and `fs-sim` packages from `ouroboros-network` repo.

### PR DESCRIPTION
**Pull request by jorisdral**
Part of https://github.com/input-output-hk/ouroboros-network/issues/4082.

This PR migrates the `fs-api` and `fs-sim` packages from the `ouroboros-network` repository to the current repository. The code is copied as-is from the [ouroboros-network#0659e7b7ef66b01763cdf15c2f8777a9394c248d commit](https://github.com/input-output-hk/ouroboros-network/commit/0659e7b7ef66b01763cdf15c2f8777a9394c248d) . We also add a GitHub Actions workflow file, and a script for running the `stylish-haskell` formatter.

Follow-up TODOs:
* See the follow-up TODOs in https://github.com/input-output-hk/ouroboros-network/pull/4425

### CODE REVIEW

File: .github/workflows/haskell.yml [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+on:
+  push:
+    branches: [ "main" ]
+  pull_request:
+
+permissions:
+  contents: read
+
+jobs:
```
Comment (dnadales) [inline .github/workflows/haskell.yml:11]: Not for this PR, but I wonder if we want to invest time in abstracting away this CI setup we replicate in multiple repos.

File: fs-api/fs-api.cabal
Comment (dnadales) [inline fs-api/fs-api.cabal:1]: Do we want to add a cabal formatter to our CI?

**Review by dnadales — APPROVED** [review]

File: .github/workflows/haskell.yml (hunk 1 above)
Comment (jorisdral) [reply .github/workflows/haskell.yml:11]: I will create an issue for this

File: fs-api/fs-api.cabal
Comment (jorisdral) [reply fs-api/fs-api.cabal:1]: Idem https://github.com/input-output-hk/fs-sim/pull/2#discussion_r1140351248

File: fs-api/src-unix/System/IO/FS.hs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  , open
+  , pread
+  , read
+  , sameError
+  , seek
+  , truncate
+  , write
+  ) where
+
+import           Prelude hiding (read, truncate)
```
[the remark this answers is not in the document]
Comment (jorisdral) [reply fs-api/src-unix/System/IO/FS.hs:16]: Idem https://github.com/input-output-hk/fs-sim/pull/2#discussion_r1140351248

**Pull request merged**